cpp/src/transpose/transpose.cu

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/transpose.hpp>
+#include <cudf/detail/transpose.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/utilities/nvtx_utils.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/detail/copy.hpp>
+#include <utilities/integer_utils.hpp>
+
+namespace cudf {
+
+namespace {
+
+constexpr int WARP_SIZE = 32;
+constexpr int MAX_GRID_SIZE = (1<<16)-1;
+
+/**
+ * @brief Transposes the values from ncols x nrows input to nrows x ncols output 
+ * 
+ * @tparam T  Datatype of values pointed to by the views
+ * @param input[in]  Device view of input columns' data
+ * @param output[out]  Mutable device view of pre-allocated output columns' data
+ */
+template <typename T>
+__global__
+void gpu_transpose(table_device_view const input, mutable_table_device_view output)
+{
+  size_type x = blockIdx.x * blockDim.x + threadIdx.x;
+  size_type y = blockIdx.y * blockDim.y + threadIdx.y;
+
+  size_type stride_x = blockDim.x * gridDim.x;
+  size_type stride_y = blockDim.y * gridDim.y;
+
+  for (size_type i = x; i < input.num_columns(); i += stride_x) {
+    for (size_type j = y; j < input.num_rows(); j += stride_y) {
+      output.column(j).element<T>(i) = input.column(i).element<T>(j);
+    }
+  }
+}
+
+/**
+ * @brief Transposes the null mask from ncols x nrows input to nrows x ncols output
+ * 
+ * @tparam T  Datatype of values pointed to by the views
+ * @param input[in]  Device view of input columns' data
+ * @param output[out]  Mutable device view of pre-allocated output columns' data
+ */
+__global__
+void gpu_transpose_valids(table_device_view const input, mutable_table_device_view output)
```

Comment:
Should add computing the null count for each output column.

Reply:
Will do